File: Week_04/515/find_largest_value_in_each_tree_row2.go
package main

func largestValues2(root *TreeNode) []int {
	res := make([]int, 0)

	var dfs func(node *TreeNode, level int)
	dfs = func(node *TreeNode, level int) {
		if node == nil {
			return
		}
		if len(res) == level {
			res = append(res, node.Val)
		} else if node.Val > res[level] {
			res[level] = node.Val
		}

		if node.Left != nil {
			dfs(node.Left, level+1)
		}
		if node.Right != nil {
			dfs(node.Right, level+1)
		}
	}
	dfs(root, 0)
	return res
}

File: Week_04/515/find_largest_value_in_each_tree_row2.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        res = []

        def dfs(node: True, level: int):
            if node is None:
                return
            if level == len(res):
                res.append(node.val)
            elif node.val > res[level]:
                res[level] = node.val

            dfs(node.left, level + 1)
            dfs(node.right, level + 1)

        dfs(root, 0)
        return res
